# Difference

The `git diff` command allows you to chech difference between condition of repository.

## Untracked files

Untracked files wont be displayed in `git diff`.

The following example just init git repository and add a file there and the `git diff` command doesn't print anything in this case.

In [2]:
%%bash
mkdir diff_example
cd diff_example

git init &> /dev/null
echo "some text" > test_file
git diff

cd ..
rm -r diff_example

## Changed files

By default `git diff` shows unstaged changes. In fact, any change to the file can be described as an insert/delete. So the following subsections show what it's like to have insert/delete lines in the `git diff` results.

### Insert line

Main point of the next cell is to add line `"second line"` to the file commited before. In `git diff` results it'll look like `+second line`.

In [5]:
%%bash
mkdir diff_example
cd diff_example

git init &> /dev/null
echo "first line" > test_file
git add test_file &> /dev/null
git commit -m "initial commit" &> /dev/null


echo "second line" >> test_file
git diff

cd ..
rm -r diff_example

diff --git a/test_file b/test_file
index 08fe272..06fcdd7 100644
--- a/test_file
+++ b/test_file
@@ -1 +1,2 @@
 first line
+second line


### Delete line

In this case I created files with two lines `"first line"` and `"second line"`. After committing, I save the new version of the file without the `second line`. Gid diff shows it as `-second line`.

In [12]:
%%bash
mkdir diff_example
cd diff_example
git init &> /dev/null

cat > test_file << EOF
first line
second line
EOF
git add test_file
git commit -m "initial commit" &> /dev/null

cat > test_file << EOF
first line
EOF
git diff

cd ..
rm -r diff_example

diff --git a/test_file b/test_file
index 06fcdd7..08fe272 100644
--- a/test_file
+++ b/test_file
@@ -1,2 +1 @@
 first line
-second line


## New file

In this section, I'll show you what a newly added file looks like in git diff.

### Empty file

In the following example, I've added an empty `second_file` and compared it with the state of the repository in the previous commit, if there is no `second_file' line, the `new file mode` will signal that this file has been created.

In [20]:
%%bash
mkdir diff_example
cd diff_example
git init &> /dev/null

touch first_file
git add first_file
git commit -m "first commit" &> /dev/null

touch second_file
git add second_file &> /dev/null

git diff --staged
cat second_file

cd ..
rm -r diff_example

diff --git a/second_file b/second_file
new file mode 100644
index 0000000..e69de29


### File with content

In this example I have created a file `second_file` and added some content to it. It'll share all the details of adding empty file option - `new file mode 100644` signal that new file has been created. But there is also a section with changes to the file below. Note that the source file from the commit to which we are comparing the state is marked as `-- /dev/null` (in Linux this directory is used for information that should simply be deleted).

In [21]:
%%bash
mkdir diff_example
cd diff_example
git init &> /dev/null

touch first_file
git add first_file
git commit -m "first commit" &> /dev/null

echo "some content" > second_file
git add second_file &> /dev/null

git diff --staged
cat second_file

cd ..
rm -r diff_example

diff --git a/second_file b/second_file
new file mode 100644
index 0000000..2ef267e
--- /dev/null
+++ b/second_file
@@ -0,0 +1 @@
+some content
some content


## --staged option

By default `git diff` shows only changed but not staged files. To get information about staged files use `--staged` option.

### New file

By using the `--staged` option, you can even see the difference between a newly created (but added) file.

So in the following example I have created and added to the repo file and compared the results with and without the `--staged` option.

In [11]:
%%bash
mkdir diff_example
cd diff_example

git init &> /dev/null
echo "some text" > staged_file
git add staged_file

echo "=====just diff===="
git diff

echo
echo "=====diff staged====="
git diff --staged

cd ..
rm -r diff_example

=====just diff====

=====diff staged=====
diff --git a/staged_file b/staged_file
new file mode 100644
index 0000000..7b57bd2
--- /dev/null
+++ b/staged_file
@@ -0,0 +1 @@
+some text


### Unstaged changes

If you use the `--staged` option, you won't see any difference when unstaging files. In the example:

- In the first commit I just added `first_file`;
- I added `second_file` and staged it;
- I changed `first_file` but did not stage it;
- Using `git diff' without the `--staged' option I only got changes in `first_file` because it wasn't staged;
- Using `git diff --staged` I only got information about `second_file` but not about `first_file`.

In [14]:
%%bash
mkdir diff_example
cd diff_example

git init &> /dev/null

echo "first text" > first_file
git add first_file
git commit -m "first file add" &> /dev/null

echo "sefond text" > second_file
git add second_file
echo "changed first text" > first_file

echo "=====just diff===="
git diff

echo
echo "=====diff staged====="
git diff --staged

cd ..
rm -r diff_example

=====just diff====
diff --git a/first_file b/first_file
index 38181e5..fc9cf92 100644
--- a/first_file
+++ b/first_file
@@ -1 +1 @@
-first text
+changed first text

=====diff staged=====
diff --git a/second_file b/second_file
new file mode 100644
index 0000000..422845a
--- /dev/null
+++ b/second_file
@@ -0,0 +1 @@
+sefond text
